### ISSUE TITLE
NaNs in relevance output

### ISSUE THREAD

**Issue by gsarti** [rachtibat/LRP-eXplains-Transformers#29]
Hi @rachtibat,

I started playing with the new efficient module, but by applying what is suggested in the quickstart tutorial the moment I am getting `nans` in the relevance tensor before normalization. What could be the issue?

## To reproduce

```python
from transformers.models.qwen2 import modeling_qwen2
import torch
from lxt.utils import pdf_heatmap, clean_tokens
from lxt.efficient import monkey_patch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

monkey_patch(modeling_qwen2, verbose=True)

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-1.5B-Instruct")
model = modeling_qwen2.Qwen2ForCausalLM.from_pretrained(
    "Qwen/Qwen2.5-1.5B-Instruct",
    quantization_config=BitsAndBytesConfig(load_in_8bit=True),
    device_map="cuda"
)

messages = [
    {"role": "system", "content": "You are a helpful assistant. Answer the user questions by reasoning step-by-step. Reasoning step MUST be given in a numbered list, while the final answer must be preceded by 'Answer:'."},
    {"role": "user", "content": "There are 4 apples on the table. Jack eats one. Elen eats a pear, and adds an apple to the table. How many apples are on the table?"},
    {"role": "assistant", "content": "Let's think step by step:"},
]

chat_str = tokenizer.apply_chat_template(messages, tokenize=False, continue_final_message=True)

for param in model.parameters():
    param.requires_grad = False

# Optionally enable gradient checkpointing (2x forward pass)
model.train()
model.gradient_checkpointing_enable()

# Get input embeddings
input_ids = tokenizer(chat_str, return_tensors="pt", add_special_tokens=True).input_ids.to(model.device)
input_embeds = model.get_input_embeddings()(input_ids)

# Inference
output_logits = model(inputs_embeds=input_embeds.requires_grad_(), use_cache=False).logits

# Take the maximum logit at last token position. You can also explain any other token, or several tokens together!
max_logits, max_indices = torch.max(output_logits[0, -1, :], dim=-1)

# Backward pass (the relevance is initialized with the value of max_logits)
max_logits.backward()

# Obtain relevance. (Works at any layer in the model!)
relevance = (input_embeds.grad * input_embeds).float().sum(-1).detach().cpu()  # Cast to float32 for higher precision

relevance
# > tensor([[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
#         nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
#         nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
#         nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]])
```

**Comment by gsarti** [comment]
Makes sense, I confirm it works when adding `torch_dtype=torch.bfloat16`! :)
